lib/fiddle/ffi_backend.rb

```diff
@@ -159,12 +159,14 @@ def call(*args, &block)
           args[i] = Fiddle::FFIBackend.to_ffi_type(args[i])
         end
       else
-        args.map! do |arg|
+        args.map!.with_index do |arg, i|
```

Comment:
Can we use `zip` here?

```suggestion
        args.zip(@args).map! do |arg, type|
```

Reply:
No, because then `map!` wouldn't modify the original `args`. I tried `map!.zip` but that produces an array of `nil`. We can use `zip` if we do `args = args.zip` but I figured modifying the args in place would be more performant

Reply:
Ok, I've just rewritten the whole thing. I modelled it after Fiddle `function_call`: when the ctype is VOIDP just pass everything into `Fiddle::Pointer.[]`, except `nil`, `String`, `FFI::AbstractMemory` and `FFI::Struct` which can all be handled directly by FFI.